builder/assignmentEnvImageBuilder.go

```diff
@@ -0,0 +1,364 @@
+		return err
+	}
+	return nil
+}
+
+// build a docker image for the given language. If the image is already present,
+// then it simply pull the image.
+func (asgmtEnv *assignmentEnvironmentImageBuilder) build() error {
+
+	if !asgmtEnv.ImageExists {
+		backgroundContext := context.Background()
+		dockerClient, err := client.NewEnvClient()
+		if err != nil {
+			return errors.Wrap(err, "error in creating a docker client")
+		}
+
+		// Create a build context tar for the image.
+		// build Context is the current working directory and where the Dockerfile is assumed to be located.
+		// [cite: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/].
+		dockerfileLoc := filepath.Base(asgmtEnv.ImgBuildConfig.dockerfileLoc)
+
+		dockerBuildContext, err := asgmtEnv.ImgBuildConfig.getDockerBuildContextTar()
+		if err != nil {
+			return err
+		}
+
+		response, err := dockerClient.ImageBuild(
+			backgroundContext,
+			dockerBuildContext,
+			types.ImageBuildOptions{
+				Dockerfile: dockerfileLoc,
+				Tags:       []string{asgmtEnv.ImgBuildConfig.imageTag}})
+		if err != nil {
+			return errors.Wrap(err, "error in building docker image")
+		}
+		defer func() {
+			err := response.Body.Close()
+			if err != nil {
+				log.Println(err)
+				return
+			}
+		}()
+
+		_, err = io.Copy(os.Stdout, response.Body)
+		if err != nil {
+			return errors.Wrap(err, "error in reading image build response")
+		}
+
+		return err
+	} else {
+		return asgmtEnv.pullImage()
+	}
+}
+
+// publishImage pushes the built image to docker hub, if required, if it is not already present.
+func (asgmtEnv *assignmentEnvironmentImageBuilder) publishImage() error {
+	if !asgmtEnv.ImageExists && asgmtEnv.ImgBuildConfig.publishImage {
+		backgroundContext := context.Background()
+		dockerClient, err := client.NewEnvClient()
+		if err != nil {
+			return errors.Wrap(err, "error in creating new docker client")
+		}
+
+		authConfig := types.AuthConfig{
+			Username: asgmtEnv.ImgBuildConfig.authData.Username,
+			Password: asgmtEnv.ImgBuildConfig.authData.Password,
+		}
+		authJson, err := json.Marshal(authConfig)
+		if err != nil {
+			return errors.Wrap(err, "error in encoding authConfig")
+		}
+
+		authString := base64.URLEncoding.EncodeToString(authJson)
+
+		imageString := fmt.Sprintf("%s/%s", constants.DockerIO, asgmtEnv.ImgBuildConfig.imageTag)
+
+		response, err := dockerClient.ImagePush(backgroundContext, imageString, types.ImagePushOptions{
+			RegistryAuth: authString,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error in pushing image to hub")
+		}
+		_, err = io.Copy(os.Stdout, response)
+		if err != nil {
+			return errors.Wrap(err, "error in reading image push response")
+		}
+		defer func() {
+			err = response.Close()
+			if err != nil {
+				log.Println(err)
+				return
+			}
+		}()
+	}
+	return nil
+}
+
+// pullImage pulls the required docker image for given language from docker hub.
```

Comment:
again is this pulling the image for the language or for the entire assignment environment?

Reply:
Based on the answer, update the comment for build().

Reply:
Yes. For assignment environment